cou/apps/base.py

```diff
@@ -519,57 +528,76 @@ def _get_upgrade_current_release_packages_step(
-        makes the migration.
-        :param target: OpenStack release as target to upgrade.
+        :param target: OpenStack release as target to upgrade
         :type target: OpenStackRelease
         :raises ApplicationError: When application has unexpected channel.
-        :return: Step for refreshing the charm.
+        :return: Step for refreshing the charm
         :rtype: PreUpgradeStep
         """
-        if not self.can_upgrade_to:
-            return PreUpgradeStep()
+        if self.is_from_charm_store:
+            return self._get_charmhub_migration_step()
+        if self.channel == LATEST_STABLE:
+            return self._get_change_to_openstack_channels_step()
+        if self._need_to_refresh(target):
+            return self._get_refresh_current_channel_step()
+        return PreUpgradeStep()
 
-        switch = None
-        *_, channel = self.possible_current_channels
+    def _get_charmhub_migration_step(self) -> PreUpgradeStep:
+        """Get the step for charm hub migration from charm store if necessary.
 
-        # corner case for rabbitmq and hacluster.
-        if len(self.possible_current_channels) > 1:
-            logger.info(
-                "'%s' has more than one channel compatible with the current OpenStack release: "
-                "'%s'. '%s' will be used",
-                self.name,
-                self.current_os_release.codename,
-                channel,
-            )
+        :return: Step for charmhub migration
+        :rtype: PreUpgradeStep
+        """
+        return PreUpgradeStep(
+            f"Migrate '{self.name}' from charmstore to charmhub",
+            coro=self.model.upgrade_charm(
+                self.name, self.expected_current_channel, switch=f"ch:{self.charm}"
+            ),
+        )
 
-        if self.origin == "cs":
-            description = f"Migrate '{self.name}' from charmstore to charmhub"
-            switch = f"ch:{self.charm}"
-        elif self.channel in self.possible_current_channels:
-            channel = self.channel
-            description = f"Refresh '{self.name}' to the latest revision of '{channel}'"
-        elif self.channel_codename >= target:
-            logger.info(
-                "Skipping charm refresh for %s, its channel is already set to %s.",
-                self.name,
-                self.channel,
-            )
-            return PreUpgradeStep()
-        elif self.channel not in self.possible_current_channels:
-            raise ApplicationError(
-                f"'{self.name}' has unexpected channel: '{self.channel}' for the current workload "
-                f"version and OpenStack release: '{self.current_os_release.codename}'. "
-                f"Possible channels are: {','.join(self.possible_current_channels)}"
-            )
+    def _get_change_to_openstack_channels_step(self) -> PreUpgradeStep:
+        """Get the step for changing to OpenStack channels if necessary.
```

Comment:
Now that we moved the conditional logic outside the function, we should update the docstring: 
```suggestion
        """Get the step for changing to OpenStack channels.
```